src/robot/util.cpp

```diff
@@ -217,49 +223,70 @@ trajectory::TrajectoryPtr planToTSR(
 
   auto robot = metaSkeleton->getBodyNode(0)->getSkeleton();
   SnapConfigurationToConfigurationPlanner::Result pResult;
-
-  auto planner = std::make_shared<SnapConfigurationToConfigurationPlanner>(
+  auto snapPlanner = std::make_shared<SnapConfigurationToConfigurationPlanner>(
       space, std::make_shared<GeodesicInterpolator>(space));
+
+  std::vector<MetaSkeletonStateSpace::ScopedState> configurations;
+
+  // Use a ranker
+  ConfigurationRankerPtr configurationRanker(ranker);
+  if (!ranker)
+  {
+    auto nominalState = space->createState();
+    space->copyState(startState, nominalState);
+    configurationRanker = std::make_shared<NominalConfigurationRanker>(
+        space, metaSkeleton, std::vector<double>(), nominalState);
+  }
+
   while (snapSamples < maxSnapSamples && generator->canSample())
   {
     // Sample from TSR
-    {
-      std::lock_guard<std::mutex> lock(robot->getMutex());
-      bool sampled = generator->sample(goalState);
-      if (!sampled)
-        continue;
+    std::lock_guard<std::mutex> lock(robot->getMutex());
+    bool sampled = generator->sample(goalState);
 
-      // Set to start state
-      space->setState(metaSkeleton.get(), startState);
-    }
+    // Increment even if it's not a valid sample since this loop
+    // has to terminate even if none are valid.
     ++snapSamples;
 
-    // Create ConfigurationToConfiguration Problem.
-    // NOTE: This is done here because the ConfigurationToConfiguration
-    // problem stores a *cloned* scoped state of the passed state.
+    if (!sampled)
+      continue;
+
+    configurations.emplace_back(goalState.clone());
+  }
+
+  {
+    // Set to start state
+    std::lock_guard<std::mutex> lock(robot->getMutex());
+    space->setState(metaSkeleton.get(), startState);
+  }
+
+  if (configurations.size() == 0)
+    return nullptr;
+
+  configurationRanker->rankConfigurations(configurations);
+
+  // Try snap planner first
+  for (std::size_t i = 0; i < configurations.size(); ++i)
+  {
     auto problem = ConfigurationToConfiguration(
-        space, startState, goalState, collisionTestable);
-    auto traj = planner->plan(problem, &pResult);
+        space, startState, configurations[i], collisionTestable);
+
+    auto traj = snapPlanner->plan(problem, &pResult);
 
     if (traj)
+    {
+      space->setState(metaSkeleton.get(), startState);
```

Comment:
Is this necessary? If we do `plan` properly, shouldn't it restore the metaskeleton state?

Reply:
Looking at the code, looks like it won't happen. 

This is probably because we are not using `planner::dart::ConfigurationToConfiguration` problem and rightfully so since we are not using a Dart planner, so the `problem` does not hold the metaskeleton for the planner to set the skeleton back. Alternatively, we can 
1. use the adapter `ConfigurationToConfiguration_to_ConfigurationToConfiguration` (let's rename this soon, it's painful to even look at :P) to make the snap planner hold the skeleton
2. In the adapter instead of `return mDelegate->plan()`, we set the skeleton back to original state and return the planned trajectory.

Reply:
Whoops, I think we're both wrong. We've already created a `MetaSkeletonStateSaver` so this should already be taken care of?

https://github.com/personalrobotics/aikido/blob/fac84625aa8fc1a6893a1d581d29dc0a3c52c81d/src/robot/util.cpp#L217

Reply:
Ah, yeah I think @brianhou 's last comment is correct. Let me remove that line.